### DISCUSSION TITLE
What is the best way to extend BaseUser

### DISCUSSION THREAD

**Discussion by jrocketdev** [piccolo-orm/piccolo#420] [Q&A]
Hopefully the question is self-explanatory enough. I'm new to piccolo, coming from mainly using sqlalchemy previously. I love the 'battery included' BaseUser class and 'user' app. However, I'd like to build in multifactor authentication and would like to add a phone number column to the BaseUser class. What is the most appropriate way to do that while also still taking advantage of the nice 'user' app commands? I can't think of an elegant approach to make sure the migration get updated correctly and I don't break any of the user' app features. I've seen a suggestion to just create a new table with a foreign key but would rather modify the initial table instead for this other column that I will always consider mandatory. Thanks for any help!

**Comment by dantownsend** [comment]
@rogersjw In my own apps I usually create a `Profile` table with a `ForeignKey` to `BaseUser`, which includes any custom fields.

```python
class Profile(Table):
    user = ForeignKey(BaseUser)
    phone_number = Varchar()
```

There isn't an official way at the moment to add fields to the `BaseUser` table.

You could copy the entire user app ([see source](https://github.com/piccolo-orm/piccolo/tree/master/piccolo/apps/user)) from Piccolo, and create your own custom app based off it.

If you're using Piccolo Admin, then `create_admin` accepts an `auth_table` argument ([see docs](https://piccolo-admin.readthedocs.io/en/latest/asgi/index.html#source)) - you could tell it to use your custom `BaseUser` table, instead of the default one.

I hope that's helpful.

**Reply by sinisaos** [reply]
@dantownsend Both methods work. Can you add a section on extending `BaseUser` to docs? This would be useful to other users.

**Reply by dantownsend** [reply]
@sinisaos Yeah, you're right - this should be added to the docs.
